relayer/relayer/src/redis/mod.rs

```diff
@@ -0,0 +1,161 @@
+use std::sync::{Arc, Mutex};
+
+use log::{debug, info, warn};
+use redis::{Client as RedisClient, Commands, Connection, RedisError};
+use thiserror::Error;
+use tokio::{
+    select,
+    sync::{broadcast, mpsc},
+};
+
+use crate::{config::Config, CircuitBreakerEvent};
+
+pub const ETH_BLOCK_KEY: &str = "ethereum_block_number";
+pub const ALEPH_BLOCK_KEY: &str = "alephzero_block_number";
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+#[non_exhaustive]
+pub enum RedisManagerError {
+    #[error("redis connection error")]
+    Redis(#[from] RedisError),
+
+    #[error("channel send error")]
+    Send(#[from] mpsc::error::SendError<u32>),
+
+    #[error("channel broadcast error")]
+    Broadcast(#[from] broadcast::error::SendError<u32>),
+
+    #[error("channel receive error")]
+    Receive(#[from] broadcast::error::RecvError),
+}
+
+pub struct RedisManager;
+
+impl RedisManager {
+    pub async fn run(
+        config: Arc<Config>,
+        next_unprocessed_block_number_eth: broadcast::Sender<u32>,
+        mut last_processed_block_number_eth: broadcast::Receiver<u32>,
+        next_unprocessed_block_number_azero: broadcast::Sender<u32>,
+        mut block_seal_receiver_azero: mpsc::Receiver<u32>,
+        mut circuit_breaker_receiver: broadcast::Receiver<CircuitBreakerEvent>,
+    ) -> Result<CircuitBreakerEvent, RedisManagerError> {
+        let Config {
+            redis_node,
+            name,
+            default_sync_from_block_eth,
+            default_sync_from_block_azero,
+            override_azero_cache,
+            override_eth_cache,
+            ..
+        } = &*config;
+
+        let client = RedisClient::open(redis_node.clone())?;
+        let redis_connection = Arc::new(Mutex::new(client.get_connection()?));
+
+        if *override_azero_cache {
+            write_block_number(
+                config.name.clone(),
+                ALEPH_BLOCK_KEY.to_string(),
+                redis_connection.clone(),
+                *config.default_sync_from_block_azero,
+            )?;
+        }
+
+        if *override_eth_cache {
+            write_block_number(
+                config.name.clone(),
+                ETH_BLOCK_KEY.to_string(),
+                redis_connection.clone(),
+                *config.default_sync_from_block_eth,
+            )?;
+        }
```

Comment:
This should not be done here, when it is and override_cache is specified, then it will reset the cache after each circuit break (or we need some "global" to guard for re-execution)

Reply:
[https://github.com/Cardinal-Cryptography/most/pull/131/commits/9f2c5d46e2412104191ec1d74019d79336138e84](https://github.com/Cardinal-Cryptography/most/pull/131/commits/9f2c5d46e2412104191ec1d74019d79336138e84)